haystack/components/routers/metadata_router.py

```diff
@@ -73,38 +96,64 @@ def __init__(self, rules: dict[str, dict]):
             },
             }
             ```
+            :param output_type: The type of the output produced. Lists of Documents or ByteStreams can be specified.
         """
         self.rules = rules
+        self.output_type = output_type
         for rule in self.rules.values():
             if "operator" not in rule:
                 raise ValueError(
                     "Invalid filter syntax. See https://docs.haystack.deepset.ai/docs/metadata-filtering for details."
                 )
-        component.set_output_types(self, unmatched=list[Document], **dict.fromkeys(rules, list[Document]))
+        component.set_output_types(self, unmatched=self.output_type, **dict.fromkeys(rules, self.output_type))
 
-    def run(self, documents: list[Document]):
+    def run(self, documents: list[Union[Document, ByteStream]]) -> dict[str, list[Union[Document, ByteStream]]]:
```

Comment:
I wonder if doing this would be better? I think technically we could handle both mixed in one list but probably better to encourage one type or the other.
```suggestion
    def run(self, documents: Union[list[Document], list[ByteStream]]) -> dict[str, Union[list[Document], list[ByteStream]]]:
```

Reply:
> To keep it this way we'll need a workaround in typing of filter methods.

We can use overloads for this. See for example here https://github.com/deepset-ai/haystack/blob/5f3c37d287a4fa882d59161a20f762e9102bc6aa/haystack/dataclasses/streaming_chunk.py#L189-L205

Overloads allow us to define the signature of a method multiple times to help with this situation. It basically allows us to say if a Document comes in as input then a Document will come out as output

Reply:
Okay I can look at that. I thought we could also use some thing like:
`T = TypeVar('T', Document, ByteStream)
`

But what's the downside of using the typing for mixed types. Also asking for my understanding.